### DISCUSSION TITLE
Running the backend separately from the frontend

### DISCUSSION THREAD

**Discussion by wnleao** [sevdokimov/log-viewer#36] [General]
How are you running the project during development? Are you always building and running the backend and frontend altogether? I would like to know if you want to decouple the project so that you are able to run each part on its own. I'm able to run the frontend with ng serve. But, right now, I don't know how to run only the backend without serving the webpage in, for instance, localhost:8111.

**Comment by sevdokimov** [comment]
I run `ng build --watch` in `log-viewer/log-viewer/src/main/frontend` folder. It build UI stuff to `log-viewer/log-viewer/target/classes/log-viewer-web` . Then I run the java app, the javaapp's classpath contains UI stuff as a regular resource. 
When I do any change in UI, `ng build --watch` compiles it immediately and java starts using the new version without restart.

Why do you want to run it separately? 
There is some problem with running UI with "ng serve" - currently backend hosts UI stuff using LogViewerServlet servlet, it patches index.html , "ng serve" will not patch it. LogViewerServlet inserts right path into `<base href="$PATH"/>`. When log-viewer is configured as embedded library in a Java Web project, the user may set any URI for LogViewerServlet servlet, so `<base href="$PATH"/>` must have the same URI. Maybe there is a more elegant solution than patching index.html, I don't know, I'm not a professional in UI development.

**Reply by wnleao** [reply]
I now have a better understandment of the way you're running the whole solution while testing and developing it. I will close my current pull request for now in order to review it if it still works in the way you are running them.

> Why do you want to run it separately?

Usually when working with fullstack solutions I tend to build things so that they can be run and tested independenty, that helps, for instance, people that are more comfortable with backend to focus on doing things there testing only the comms interface without having to run the frontend aswell and vice-versa. Other example, I used "ng serve" and tried to access the default localhost:4200 to focus on learning the frontend part and I got a "Cannot GET /rest/navigator/initState", then I went back to running the LogViewerServlet. It's not a problem, it's more of a question of understanding how things were built before tinkering with them.

> LogViewerServlet inserts right path into <base href="$PATH"/>. When log-viewer is configured as embedded library in a Java Web project, the user may set any URI for LogViewerServlet servlet, so <base href="$PATH"/> must have the same URI. 

After I review the pull request I'll take a look into that and see what we could do.

Also, I would like to state that this is a great project! I would be glad to keep helping.

**Reply by wnleao** [reply]
I've tested the frontend using `ng build --watch` and submitted a new pull request.
